cpp/include/cudf/binning/bin.cuh

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// TODO: Clean up includes before pushing a final version.
+#include <cudf/column/column_view.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/binning/bin.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/types.hpp>
+#include <thrust/functional.h>
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <cudf/copying.hpp>
+
+namespace cudf {
+
+namespace bin {
+
+// TODO: This is a placeholder to remind myself to figure out the proper way to
+// set nulls. I need to modify the bitmask, but even if I passed a pointer to
+// it directly to the constructor of bin_finder and modified it directly (which
+// would be a terribly hacky solution), the current index is not provided to
+// the call operator so I wouldn't know what bit to set (and in any case that
+// would not be parallel-safe without atomics).
+constexpr unsigned int MYNULL = 0xffffffff;
+
+namespace detail {
+namespace {
```

Comment:
Don't use anonymous namespaces in headers. https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file

I don't think any of this needs to be in a header. I'd put all of this code inside `bin.cu` inside an anon namespace.

Reply:
You're OK with putting templates into the source file, I assume because in this case we can safely assume that none of these will be used in any other translation unit?

Reply:
Right, if the visibility is only limited to that TU with an anon namespace, then it's not possible for it to be used elsewhere.

Reply:
I've done this, just leaving the thread since I'm curious to hear your thoughts re: templates in source files, since I usually avoid that.

Reply:
I put templates in source files all the time if I'm only ever going to use that template in that TU.